File: ffi/secretary-ffi-uniffi/tests/swift/SmokeFolderIn.swift
// Folder-in open assertions for the Swift smoke runner.
//
// Covers:
//   - B.4a: open_vault_with_password (success + wrong-password + nonexistent)
//   - Issue #30 follow-up: open_vault_with_recovery (success + 3-word phrase
//     + wrong-vault phrase)
//
// Folder-in entry points take a vault folder path on disk and load
// vault.toml + identity bundle internally. The bytes-in counterpart
// lives in SmokeBytesIn.swift.

import Foundation

func runFolderInAsserts(env: SmokeEnv) {
    // =============================================================================
    // B.4a — folder-in open_vault_with_password asserts
    // =============================================================================

    // Assert 16: open_vault_with_password success — identity + manifest both populated.
    do {
        let folderPath = Data(env.vault001Url.path.utf8)
        let out = try openVaultWithPassword(folderPath: folderPath, password: env.password001)
        defer { out.identity.wipe() }
        defer { out.manifest.wipe() }
        let displayName = out.identity.displayName()
        let blockCount = out.manifest.blockCount()
        check(
            displayName == expectedDisplayName && blockCount > 0,
            "open_vault_with_password success → displayName=\"\(displayName)\", blockCount=\(blockCount)"
        )
    } catch {
        check(false, "open_vault_with_password success threw \(error), expected to succeed")
    }

    // Assert 17: open_vault_with_password wrong password → VaultError.WrongPasswordOrCorrupt.
    do {
        let folderPath = Data(env.vault001Url.path.utf8)
        let wrongPassword = Data("definitely wrong".utf8)
        _ = try openVaultWithPassword(folderPath: folderPath, password: wrongPassword)
        check(false, "wrong password should have thrown VaultError.WrongPasswordOrCorrupt")
    } catch VaultError.WrongPasswordOrCorrupt {
        check(true, "open_vault_with_password wrong password → VaultError.WrongPasswordOrCorrupt")
    } catch {
        check(false, "wrong password (vault) threw \(error), expected VaultError.WrongPasswordOrCorrupt")
    }

    // Assert 18: nonexistent folder → VaultError.FolderInvalid with detail.
    do {
        let folderPath = Data("/tmp/__nonexistent_b4a_swift__".utf8)
        _ = try openVaultWithPassword(folderPath: folderPath, password: env.password001)
        check(false, "nonexistent folder should have thrown VaultError.FolderInvalid")
    } catch let e as VaultError {
        if case let .FolderInvalid(detail) = e {
            let lc = detail.lowercased()
            check(
                lc.contains("vault.toml") || lc.contains("no such file"),
                "nonexistent folder → VaultError.FolderInvalid(detail=\"\(detail)\")"
            )
        } else {
            check(false, "nonexistent folder threw wrong VaultError variant: \(e)")
        }
    } catch {
        check(false, "nonexistent folder threw \(error), expected VaultError.FolderInvalid")
    }

    // =============================================================================
    // Issue #30 follow-up — folder-in open_vault_with_recovery asserts
    // =============================================================================
    //
    // Mirrors asserts 16-18 (folder-in password) but exercises the recovery
    // path through the folder-in entry point. The bytes-in
    // `open_with_recovery` surface is already covered by asserts 9-12 in
    // SmokeBytesIn.swift; the folder-in `open_vault_with_recovery`
    // counterpart was missing. Pinned KAT inputs come from
    // `golden_vault_001_inputs.json` via `_phraseFromInputs`.

    // Assert 35: open_vault_with_recovery success — identity + manifest both populated.
    do {
        let folderPath = Data(env.vault001Url.path.utf8)
        let out = try openVaultWithRecovery(folderPath: folderPath, mnemonic: env.phrase001)
        defer { out.identity.wipe() }
        defer { out.manifest.wipe() }
        let displayName = out.identity.displayName()
        let blockCount = out.manifest.blockCount()
        check(
            displayName == expectedDisplayName && blockCount > 0,
            "open_vault_with_recovery success → displayName=\"\(displayName)\", blockCount=\(blockCount)"
        )
    } catch {
        check(false, "open_vault_with_recovery success threw \(error), expected to succeed")
    }

    // Assert 36: open_vault_with_recovery 3-word phrase → VaultError.InvalidMnemonic(detail).
    do {
        let folderPath = Data(env.vault001Url.path.utf8)
        let bad = Data("only three words".utf8)
        _ = try openVaultWithRecovery(folderPath: folderPath, mnemonic: bad)
        check(false, "3-word phrase should have thrown VaultError.InvalidMnemonic")
    } catch let e as VaultError {
        if case let .InvalidMnemonic(detail) = e {
            check(
                detail.contains("got 3"),
                "open_vault_with_recovery 3-word → VaultError.InvalidMnemonic(detail=\"\(detail)\") mentions `got 3`"
            )
        } else {
            check(false, "3-word phrase threw wrong VaultError variant: \(e)")
        }
    } catch {
        check(false, "3-word phrase threw \(error), expected VaultError.InvalidMnemonic")
    }

    // Assert 37: open_vault_with_recovery vault_002 phrase against vault_001 folder → WrongMnemonicOrCorrupt.
    do {
        let folderPath = Data(env.vault001Url.path.utf8)
        _ = try openVaultWithRecovery(folderPath: folderPath, mnemonic: env.phrase002)
        check(false, "vault_002 phrase against vault_001 folder should have thrown VaultError.WrongMnemonicOrCorrupt")
    } catch VaultError.WrongMnemonicOrCorrupt {
        check(true, "open_vault_with_recovery wrong-vault phrase → VaultError.WrongMnemonicOrCorrupt")
    } catch {
        check(false, "wrong-vault phrase threw \(error), expected VaultError.WrongMnemonicOrCorrupt")
    }

    // =============================================================================
    // create_vault_in_folder — write a complete vault on disk, then open it.
    // =============================================================================

    // Assert 38: create_vault_in_folder writes a complete, openable vault
    // (24-word recovery phrase + folder-password open succeeds → browsable).
    do {
        let tmp = FileManager.default.temporaryDirectory
            .appendingPathComponent("create-folder-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: tmp, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: tmp) }

        let folderPath = Data(tmp.path.utf8)
        let pw = Data("create-smoke-pw".utf8)
        // createVaultInFolder returns CreatedVaultInFolder holding the one-shot
        // MnemonicOutput handle — take the phrase from the inner handle and wipe it.
        // (This block had rotted against the pre-Slice-5 MnemonicOutput-returning
        // shape; the smoke runner is not a CI gate, so it only surfaced when re-run
        // for #307.)
        let created = try createVaultInFolder(
            folderPath: folderPath,
            password: pw,
            displayName: "Swift-Create-Bob",
            createdAtMs: 1_700_000_000_000
        )
        defer { created.mnemonic.wipe() }
        let wordCount = created.mnemonic.takePhrase().map {
            String(decoding: $0, as: UTF8.self).split(separator: " ").count
        } ?? 0

        let opened = try openVaultWithPassword(folderPath: folderPath, password: pw)
        defer { opened.identity.wipe() }
        defer { opened.manifest.wipe() }
        check(
            wordCount == 24 && opened.identity.displayName() == "Swift-Create-Bob",
            "create_vault_in_folder → 24-word phrase + openable vault (words=\(wordCount), displayName=\"\(opened.identity.displayName())\")"
        )
    } catch {
        check(false, "create_vault_in_folder smoke threw \(error), expected success")
    }

    // Assert 39: create_vault_in_folder on a non-empty folder → VaultError.VaultFolderNotEmpty.
    do {
        let tmp = FileManager.default.temporaryDirectory
            .appendingPathComponent("create-nonempty-\(UUID().uuidString)", isDirectory: true)
        try FileManager.default.createDirectory(at: tmp, withIntermediateDirectories: true)
        defer { try? FileManager.default.removeItem(at: tmp) }
        try Data("x".utf8).write(to: tmp.appendingPathComponent("junk"))

        _ = try createVaultInFolder(
            folderPath: Data(tmp.path.utf8),
            password: Data("pw".utf8),
            displayName: "X",
            createdAtMs: 1_700_000_000_000
        )
        check(false, "non-empty folder should have thrown VaultError.VaultFolderNotEmpty")
    } catch VaultError.VaultFolderNotEmpty {
        check(true, "create_vault_in_folder non-empty → VaultError.VaultFolderNotEmpty")
    } catch {
        check(false, "non-empty folder threw \(error), expected VaultError.VaultFolderNotEmpty")
    }
}

File: ffi/secretary-ffi-uniffi/tests/kotlin/SmokeFolderIn.kt
// Folder-in open assertions for the Kotlin smoke runner.
//
// Covers:
//   - B.4a: open_vault_with_password (success + wrong-password + nonexistent)
//   - Issue #30 follow-up: open_vault_with_recovery (success + 3-word phrase
//     + wrong-vault phrase)
//
// Folder-in entry points take a vault folder path on disk and load
// vault.toml + identity bundle internally. The bytes-in counterpart
// lives in SmokeBytesIn.kt.

import uniffi.secretary.VaultException
import uniffi.secretary.createVaultInFolder
import uniffi.secretary.openVaultWithPassword
import uniffi.secretary.openVaultWithRecovery

fun runFolderInAsserts(env: SmokeEnv) {
    val goldenVault001Folder = env.vault001Path.toString()

    // =============================================================================
    // B.4a — folder-in open_vault_with_password asserts
    // =============================================================================

    // Assertion 16: open_vault_with_password success — identity + manifest both populated.
    // Chained .use { } cleanup matches the bytes-in idiom (SmokeBytesIn.kt:41-56)
    // and guarantees both handles close even if displayName() / blockCount() / check()
    // throws — the prior sequential-cleanup pattern leaked both handles on exception.
    try {
        val folderPath = goldenVault001Folder.toByteArray(Charsets.UTF_8)
        val out = openVaultWithPassword(folderPath, env.password001)
        out.identity.use { identity ->
            out.manifest.use { manifest ->
                val displayName = identity.displayName()
                val blockCount = manifest.blockCount()
                check(
                    displayName == EXPECTED_DISPLAY_NAME && blockCount > 0UL,
                    "open_vault_with_password success → displayName=\"$displayName\", blockCount=$blockCount",
                )
            }
        }
    } catch (e: Throwable) {
        check(false, "open_vault_with_password success threw $e, expected to succeed")
    }

    // Assertion 17: wrong password → VaultException.WrongPasswordOrCorrupt.
    try {
        val folderPath = goldenVault001Folder.toByteArray(Charsets.UTF_8)
        val wrongPassword = "definitely wrong".toByteArray(Charsets.UTF_8)
        openVaultWithPassword(folderPath, wrongPassword)
        check(false, "wrong password (vault) should have thrown VaultException.WrongPasswordOrCorrupt")
    } catch (e: VaultException.WrongPasswordOrCorrupt) {
        check(true, "open_vault_with_password wrong password → VaultException.WrongPasswordOrCorrupt")
    } catch (e: Throwable) {
        check(false, "wrong password (vault) threw $e, expected VaultException.WrongPasswordOrCorrupt")
    }

    // Assertion 18: nonexistent folder → VaultException.FolderInvalid with detail.
    try {
        val folderPath = "/tmp/__nonexistent_b4a_kotlin__".toByteArray(Charsets.UTF_8)
        openVaultWithPassword(folderPath, env.password001)
        check(false, "nonexistent folder should have thrown VaultException.FolderInvalid")
    } catch (e: VaultException.FolderInvalid) {
        val detail = e.detail.lowercase()
        check(
            detail.contains("vault.toml") || detail.contains("no such file"),
            "nonexistent folder → VaultException.FolderInvalid(detail=\"${e.detail}\")",
        )
    } catch (e: Throwable) {
        check(false, "nonexistent folder threw $e, expected VaultException.FolderInvalid")
    }

    // =============================================================================
    // Issue #30 follow-up — folder-in open_vault_with_recovery asserts
    // =============================================================================
    //
    // Mirrors asserts 16-18 (folder-in password) but exercises the
    // recovery path through the folder-in entry point. The bytes-in
    // `open_with_recovery` surface is already covered by asserts 9-12 in
    // SmokeBytesIn.kt; the folder-in `open_vault_with_recovery`
    // counterpart was missing. Pinned KAT inputs come from
    // `golden_vault_001_inputs.json` via `phraseFromInputs`.

    // Assert 36: open_vault_with_recovery success — identity + manifest both populated.
    // Same chained-.use { } cleanup as assert 16 above — see that comment.
    try {
        val folderPath = goldenVault001Folder.toByteArray(Charsets.UTF_8)
        val out = openVaultWithRecovery(folderPath, env.phrase001)
        out.identity.use { identity ->
            out.manifest.use { manifest ->
                val displayName = identity.displayName()
                val blockCount = manifest.blockCount()
                check(
                    displayName == EXPECTED_DISPLAY_NAME && blockCount > 0UL,
                    "open_vault_with_recovery success → displayName=\"$displayName\", blockCount=$blockCount",
                )
            }
        }
    } catch (e: Throwable) {
        check(false, "open_vault_with_recovery success threw $e, expected to succeed")
    }

    // Assert 37: open_vault_with_recovery 3-word phrase → VaultException.InvalidMnemonic(detail).
    try {
        val folderPath = goldenVault001Folder.toByteArray(Charsets.UTF_8)
        val bad = "only three words".toByteArray(Charsets.UTF_8)
        openVaultWithRecovery(folderPath, bad)
        check(false, "3-word phrase should have thrown VaultException.InvalidMnemonic")
    } catch (e: VaultException.InvalidMnemonic) {
        check(
            e.detail.contains("got 3"),
            "open_vault_with_recovery 3-word → VaultException.InvalidMnemonic(detail=\"${e.detail}\") mentions `got 3`",
        )
    } catch (e: Throwable) {
        check(false, "3-word phrase threw $e, expected VaultException.InvalidMnemonic")
    }

    // Assert 38: open_vault_with_recovery vault_002 phrase against vault_001 folder → WrongMnemonicOrCorrupt.
    try {
        val folderPath = goldenVault001Folder.toByteArray(Charsets.UTF_8)
        openVaultWithRecovery(folderPath, env.phrase002)
        check(false, "vault_002 phrase against vault_001 folder should have thrown VaultException.WrongMnemonicOrCorrupt")
    } catch (e: VaultException.WrongMnemonicOrCorrupt) {
        check(true, "open_vault_with_recovery wrong-vault phrase → VaultException.WrongMnemonicOrCorrupt")
    } catch (e: Throwable) {
        check(false, "wrong-vault phrase threw $e, expected VaultException.WrongMnemonicOrCorrupt")
    }

    // =============================================================================
    // create_vault_in_folder — write a complete vault on disk, then open it.
    // =============================================================================

    // Assert 39: create_vault_in_folder writes a complete, openable vault
    // (24-word recovery phrase + folder-password open succeeds → browsable).
    run {
        val tmp = kotlin.io.path.createTempDirectory("create-folder-").toFile()
        try {
            val folderPath = tmp.path.toByteArray(Charsets.UTF_8)
            val pw = "create-smoke-pw".toByteArray(Charsets.UTF_8)
            var wordCount = 0
            createVaultInFolder(folderPath, pw, "Kotlin-Create-Bob", 1_700_000_000_000UL).use { mn ->
                val phrase = mn.takePhrase()
                check(phrase != null, "create_vault_in_folder take_phrase returned null")
                if (phrase != null) {
                    // takePhrase() is `bytes?` → a ByteArray? directly (#261); decode UTF-8 to count words.
                    wordCount = phrase.toString(Charsets.UTF_8).split(" ").size
                }
            }
            val out = openVaultWithPassword(folderPath, pw)
            out.identity.use { identity ->
                out.manifest.use { _ ->
                    val name = identity.displayName()
                    check(
                        wordCount == 24 && name == "Kotlin-Create-Bob",
                        "create_vault_in_folder → 24-word phrase + openable vault (words=$wordCount, displayName=\"$name\")",
                    )
                }
            }
        } catch (e: Throwable) {
            check(false, "create_vault_in_folder smoke threw $e, expected success")
        } finally {
            tmp.deleteRecursively()
        }
    }

    // Assert 40: create_vault_in_folder on a non-empty folder → VaultException.VaultFolderNotEmpty.
    run {
        val tmp = kotlin.io.path.createTempDirectory("create-nonempty-").toFile()
        try {
            java.io.File(tmp, "junk").writeText("x")
            createVaultInFolder(
                tmp.path.toByteArray(Charsets.UTF_8),
                "pw".toByteArray(Charsets.UTF_8),
                "X",
                1_700_000_000_000UL,
            )
            check(false, "non-empty folder should have thrown VaultException.VaultFolderNotEmpty")
        } catch (e: VaultException.VaultFolderNotEmpty) {
            check(true, "create_vault_in_folder non-empty → VaultException.VaultFolderNotEmpty")
        } catch (e: Throwable) {
            check(false, "non-empty folder threw $e, expected VaultException.VaultFolderNotEmpty")
        } finally {
            tmp.deleteRecursively()
        }
    }
}
